python_onnxifi/data_conversion.cc

```diff
@@ -0,0 +1,298 @@
+      op_template<double, double>(__VA_ARGS__);                              \
+      break;                                                                 \
+    }                                                                        \
+    case ONNX_NAMESPACE::TensorProto_DataType_COMPLEX128: {                  \
+      op_template<std::complex<double>, std::complex<double> >(__VA_ARGS__); \
+      break;                                                                 \
+    }                                                                        \
+    case ONNX_NAMESPACE::TensorProto_DataType_STRING:                        \
+    case ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED: {                   \
+      throw std::runtime_error("Dispatch received non-numeric data type");   \
+    }                                                                        \
+  }
+
+template <typename onnx_type, typename unused>
+void DataConversion::getBufferSize(uint64_t& buffer_size,
+                                   const uint64_t* shape,
+                                   uint32_t dimensions) {
+  auto numElements = std::accumulate(shape, shape + dimensions, (uint64_t)1,
+                                     std::multiplies<uint64_t>());
+  buffer_size = sizeof(onnx_type) * numElements;
+}
+
+void DataConversion::updateOutputDescriptorsData(ModelProto& model) {
+  output_descriptors_data_.clear();
+  ::ONNX_NAMESPACE::shape_inference::InferShapes(model);
+  const auto& outputs = model.graph().output();
+  for (const auto& vi : outputs) {
+    DescriptorData outputData;
+    outputData.name = vi.name();
+    outputData.descriptor.name = outputData.name.c_str();
+    if (!vi.type().tensor_type().has_elem_type()) {
+      throw std::runtime_error(
+          "Non-static ModelProto: Output data type not found");
+    }
+    outputData.descriptor.dataType = vi.type().tensor_type().elem_type();
+    outputData.descriptor.dimensions =
+        vi.type().tensor_type().shape().dim_size();
+    ;
+    for (auto i = 0; i < outputData.descriptor.dimensions; ++i) {
+      const auto& dim = vi.type().tensor_type().shape().dim(i);
+      if (!dim.has_dim_value()) {
+        throw std::runtime_error(
+            "Non-static ModelProto: Output shape dimension not found");
+      }
+      outputData.shape.push_back(dim.dim_value());
+    }
+    outputData.descriptor.shape = outputData.shape.data();
+    outputData.descriptor.memoryType =
+        ONNXIFI_MEMORY_TYPE_CPU;  // TODO: Expand memory types?
+    uint64_t buffer_size;
+    DISPATCH_OVER_NUMERIC_DATA_TYPE(
+        outputData.descriptor.dataType, getBufferSize, buffer_size,
+        outputData.descriptor.shape, outputData.descriptor.dimensions)
+    outputData.buffer.resize(buffer_size);
+    outputData.descriptor.buffer =
+        reinterpret_cast<onnxPointer>(outputData.buffer.data());
+    output_descriptors_data_.push_back(std::move(outputData));
```

Comment:
`push_back` will still incur a copy. You need `emplace_back`

Reply:
Good point. Though std::move will prevent large memory copying, emplace_back is even better